apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemService.java

```diff
@@ -207,6 +224,116 @@ private NotFoundException failWithNotFound(String message) {
+     * @return a Mono that completes when all items have been enriched
+     */
+    private Mono<Void> enrichItemsWithTraceMetadata(String workspaceId, List<DatasetItem> items) {
+        // Get unique trace IDs from items
+        Set<UUID> traceIds = items.stream()
+                .map(DatasetItem::traceId)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+
+        log.debug("Found '{}' unique trace IDs to enrich", traceIds.size());
+
+        if (traceIds.isEmpty()) {
+            log.warn("No trace IDs found in dataset items, skipping metadata enrichment");
+            return Mono.empty();
+        }
+
+        // Fetch trace details for all traces
+        return Flux.fromIterable(traceIds)
+                .flatMap(traceId -> traceService.get(traceId)
+                        .map(trace -> {
+                            log.debug("Successfully fetched trace '{}' with metadata", traceId);
+                            return Map.entry(traceId, trace);
+                        })
+                        .onErrorResume(error -> {
+                            log.warn("Failed to fetch trace details for trace '{}': {}", traceId, error.getMessage());
+                            return Mono.empty();
+                        }))
+                .collectMap(Map.Entry::getKey, Map.Entry::getValue)
+                .doOnNext(traceMap -> {
+                    log.debug("Successfully fetched '{}' traces, enriching dataset items", traceMap.size());
+                    // Enrich items with trace metadata by replacing them in the list
+                    for (int i = 0; i < items.size(); i++) {
+                        DatasetItem item = items.get(i);
+                        if (item.traceId() != null && traceMap.containsKey(item.traceId())) {
+                            Trace trace = traceMap.get(item.traceId());
+                            DatasetItem enrichedItem = enrichItemWithTraceMetadata(item, trace);
+                            log.debug("Enriched dataset item '{}' with trace metadata", item.id());
+                            items.set(i, enrichedItem);
+                        }
+                    }
+                })
+                .then();
+    }
+
+    /**
+     * Creates a new DatasetItem with enriched data containing trace metadata.
+     * 
+     * @param item the original dataset item
+     * @param trace the trace containing metadata to add
+     * @return a new DatasetItem with enriched data
+     */
+    private DatasetItem enrichItemWithTraceMetadata(DatasetItem item, Trace trace) {
+        // Create a mutable copy of the data map
+        Map<String, JsonNode> enrichedData = new HashMap<>(item.data());
+
+        // Add trace metadata to the data map
+        if (trace.tags() != null && !trace.tags().isEmpty()) {
+            enrichedData.put("trace_tags", JsonUtils.getJsonNodeFromStringWithFallback(
```

Comment:
To reduce the risk of collision, I would create a single key called ` original_trace` and add all columns there with the same name as the trace object.